File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        stack<int> st;
        vector<int> next(n);
        for(int i=n-1;i>=0;i--){
            while(st.empty() == false && st.top() <= nums[i]){
                st.pop();
            }
            if(st.empty() == true) next[i] =-1;
            else next[i] = st.top();
            st.push(nums[i]);
        }
        for(int i=n-1;i>=0;i--){
            while(st.empty() == false && st.top() <= nums[i]){
                st.pop();
            }
            if(st.empty() == true) next[i] =-1;
            else next[i] = st.top();
            st.push(nums[i]);
        }
        return next;
    }
};

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        st =[]
        next =[-1]*n
        for i in range(n-1,-1,-1):
            while(st and st[-1] <= nums[i]):
                st.pop()
            if not st:
                next[i] = -1
            else:
                next[i] = st[-1]

            st.append(nums[i])

        for i in range(n-1,-1,-1):
            while(st and st[-1] <= nums[i]):
                st.pop()
            if not st:
                next[i] = -1
            else:
                next[i] = st[-1]
            st.append(nums[i])
        
        return next